lib/psabpf_value_set.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2022 Orange
+ * Copyright 2022 Warsaw University of Technology
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <bpf/bpf.h>
+
+#include <psabpf.h>
+#include <psabpf_value_set.h>
+#include "common.h"
+#include "btf.h"
+
+void psabpf_value_set_context_init(psabpf_value_set_context_t *ctx) {
+    if (ctx == NULL)
+        return;
+
+    memset(ctx, 0, sizeof(psabpf_value_set_context_t));
+}
+
+void psabpf_value_set_context_free(psabpf_value_set_context_t *ctx) {
+    if (ctx == NULL)
+        return;
+
+    free_btf(&ctx->btf_metadata);
+    close_object_fd(&(ctx->set_map.fd));
+    free_struct_field_descriptor_set(&ctx->fds);
+}
+
+void psabpf_value_set_init(psabpf_value_set_t *value) {
+    if (value == NULL)
+        return;
+
+    memset(value, 0, sizeof(psabpf_value_set_t));
+}
+
+void psabpf_value_set_free(psabpf_value_set_t *value) {
+    if (value == NULL)
+        return;
+
+    free_struct_field_set(&value->value);
+
+    if (value->raw_data != NULL)
+        free(value->raw_data);
+    value->raw_data = NULL;
+}
+
+static int parse_key_type(psabpf_value_set_context_t *ctx)
+{
+    uint32_t type_id = 0;
+    if (ctx->btf_metadata.btf != NULL) {
+        type_id = psabtf_get_member_type_id_by_name(ctx->btf_metadata.btf, ctx->set_map.btf_type_id, "key");
+    }
```

Comment:
```suggestion
    uint32_t type_id = psabtf_get_member_type_id_by_name(ctx->btf_metadata.btf, ctx->set_map.btf_type_id, "key");
```
`psabtf_get_member_type_id_by_name` should correctly handle NULL BTF pointer now.